### PR TITLE
Bring in merkle and tmhash from tmlibs

### PR DESCRIPTION
**Pull request by ebuchman**
Ref https://github.com/tendermint/tmlibs/pull/210#issuecomment-390995994

### CODE REVIEW

**Comment by ebuchman** [comment]
Also replaces https://github.com/tendermint/tmlibs/pull/211

**Comment by liamsi** [comment]
referencing discussion on hashing inner & leaf nodes differently; as well as the removed encoding (as both might or might not have security implications): 
https://github.com/tendermint/tmlibs/pull/211#issuecomment-391044446 (and following)

File: merkle/simple_tree.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+*/
+
+package merkle
+
+import (
+	"github.com/tendermint/go-crypto/tmhash"
+)
+
+// Hash(left | right). The basic operation of the Merkle tree.
+func SimpleHashFromTwoHashes(left, right []byte) []byte {
```
Comment (liamsi) [inline merkle/simple_tree.go]: If the size of left and right is now fixed, can we use the size in the args: 
`SimpleHashFromTwoHashes(left, right [Size]byte)`
see also: https://github.com/tendermint/tmlibs/pull/211#issuecomment-391102658

File: merkle/simple_tree.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+                 /             \
+                *               *
+               / \             / \
+              /   \           /   \
+             /     \         /     \
+            *       *       *       h6
+           / \     / \     / \
+          h0  h1  h2  h3  h4  h5
+
+*/
```
Comment (liamsi) [inline merkle/simple_tree.go]: Can we add a warning here about pre-image attacks, or, more general a warning on how to use all this safely (in our context, but also in general). See for instance: https://github.com/bitcoin/bitcoin/blob/5961b23898ee7c0af2626c46d5d70e80136578d3/src/consensus/merkle.cpp#L9-L43

Comment (liamsi) [reply merkle/simple_tree.go]: I can take care of this later too (I would open an issue though referencing relevant comments)

Comment (ebuchman) [reply merkle/simple_tree.go]: Would love it if you could pick this up! Thanks!

**Comment by ebuchman** [comment]
@Liamsi also worth noting how we use these functions from `github.com/tendermint/types`  - especially that we duplicate the implementation to avoid allocating/populating large slices. 

See PR to update tendermint dependent on this in https://github.com/tendermint/tendermint/pull/1606

**Comment by liamsi** [comment]
Thanks for the pointer! If you like, we can fix the test-vectors here and merge this (so you can reference the development branch there instead: https://github.com/tendermint/tendermint/pull/1606/files#diff-bd247e83efc3c45ae9e8c47233249f18R261). 
I'll keep track and take care of the comments (documenting safety and JP's comments).

File: merkle/simple_tree.go (hunk 2 above)
Comment (liamsi) [reply merkle/simple_tree.go]: https://github.com/tendermint/go-crypto/issues/107

File: merkle/simple_map_test.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+type strHasher string
+
+func (str strHasher) Hash() []byte {
+	return SimpleHashFromBytes([]byte(str))
```
Comment (liamsi) [inline merkle/simple_map_test.go]: Any reason SimpleHashFromBytes disappeared?
https://github.com/tendermint/tmlibs/blob/50ad19541d61a60bdd6be6f407af3ed7abd6ba44/merkle/simple_tree.go#L70-L74

File: merkle/simple_tree.go (hunk 1 above)
Comment (liamsi) [reply merkle/simple_tree.go]: Making this explicit via `left, right [tmhash.Size]byte` makes it necessary to change a lot of places. I propose to keep the encoding (called length prefix here: https://github.com/tendermint/tmlibs/pull/211) and potentially refactor this at a later point.

File: merkle/simple_map_test.go (hunk 3 above)
Comment (ebuchman) [reply merkle/simple_map_test.go]: meh - trying to simplify the exposed interface and it got replaced with `tmhash.Sum`

Comment (liamsi) [reply merkle/simple_map_test.go]: OK, makes sense

File: merkle/simple_tree.go (hunk 1 above)
Comment (ebuchman) [reply merkle/simple_tree.go]: damn ok. my motivation was to simplify the tree description by removing unecessary length prefixing. but i agree that if we do so, we should introduce some better type safety, and doing so will require ALOT of changes ... so maybe its not that important right now ? Just feels like now is a good time to make wanted breaking changes to the data structures :P

Comment (ebuchman) [reply merkle/simple_tree.go]: Also having length prefixes doesnt solve the 2nd pre-image attack right ?

Comment (liamsi) [reply merkle/simple_tree.go]: Nope. It just mitigates similar but simpler collisions (e.g., I guess where you just append 2 byte slices together and yield the same root hash as with two leafs ?). Let's  re-add the encoding (which allows for variable sized byte slices to be hashed) and add proper prefixes in a separate issue.

Comment (liamsi) [reply merkle/simple_tree.go]: > so maybe its not that important right now ? 

it's definitely easier to simply re-introduce the encoding. I've already started adding fixed lengths everywhere in another branch. There seems to be sth wrong though (didn't investigate but I pushed it now: https://github.com/Liamsi/go-crypto/compare/bucky_fixed...Liamsi:fixed_size?expand=1)

Comment (ebuchman) [reply merkle/simple_tree.go]: sth wrong how ?

Comment (ebuchman) [reply merkle/simple_tree.go]: also if we go through with https://github.com/tendermint/tmlibs/pull/213 we'll need the length prefix in here.

I guess so long as we use []byte, we should use a length prefix, at least so people don't use it wrong.

Comment (liamsi) [reply merkle/simple_tree.go]: > sth wrong how ?

TestSimpleProof fails on every index

> I guess so long as we use []byte, we should use a length prefix, at least so people don't use it wrong.

Totally.

**Comment by ebuchman** [comment]
Cool I'm going to squash some commits and then use amino for the encodeByteSlice (didn't want to do that before to avoid an amino dependency in tmlibs)

**Pull request merged**